proposals/NNNN-improve-VDE-for-stable-frame-rate.md

```diff
@@ -0,0 +1,165 @@
+                                    fullFrameRect.drawFrame(textureId, matrix);
+                                } catch(RuntimeException e) {
+                                    ...
+                                }
+                                nextTime += frameIntervalInNano;
+                            }
+                            long delayTime = // adjust delaytime..
+                            handler.sendMessageDelayed(handler.obtainMessage(MSG_TICK), delayTime);
+                            break;
+                        case MSG_UPDATE_SURFACE:
+                            // update the surface here.
+                            updateSurface();
+                    }
+                }
+         }
+    }
+
+    private void updateSurface() {
+        try {
+            windowSurface.makeCurrent()
+        } catch(RuntimeException e) {
+            ...
+        }
+        surfacetexture.updateTexImage();
+        surfaceTexture.getTransformMatrix(matrix);
+    }
+
+    /**
+    * this is where we update the surface
+    * @param surfaceTexture
+    */
+    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
+        // here, we can do update surfaceTexture
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            updateSurface();
+        } else {
+            // note that API level is lower than 21 (LOLLIPOP), setOnFrameAvailableListener(listener) is used,
+            // and most likely, OnFrameAvailableListener gets called in main thread.
+            // In that case, we have to call updateSurface in CaptureThread instead.
+            handler.sendMessage(handler.obtainMessafge(MSG_UPDATE_SURFACE));
+        }
+
+        // when the first time this gets called, start the loop
+        // by mHandler.sendMessage()
+        if (nextTime == 0) {
+            nextTime = System.nanoTime();
+            handler.sendMessage(handler.obtainMessage(MSG_TICK));
+        }
+    }
+ }
+```
+
+## Potential downsides
+
+No downside, as it will be implemented in Proxy, and there should be no performance overhead.
+One thing to note is that this proposal includes the usage of `Grafika` component, which is under [Apache License 2.0](http://www.apache.org/licenses/LICENSE-2.0).
+
+sdl_java_suite already uses some component under Apache License 2.0 (e.g. `JSON`), so adding another open source component won't cause any issues.
+
+## Impact on existing code
+
+Because this approach does not change existing API, and changes are inside of `VirtualDisplayEncoder` class, there's no impact to developers who use `VirtualDisplayEncoder`.
+
+## Alternatives considered
+
+This is pure improvement for existing `VistualDisplayEncoder`. So there're no alternatives, but this proposal adds the ability to specify the desired FPS on Proxy end.
+This proposal should be combined with "Add preferred FPS to VideoStreamingCapability" proposal, so that we can respect for preferred FPS value, which is specified by HU.
```

Comment:
```suggestion
This proposal should be combined with "Add preferred FPS to VideoStreamingCapability" proposal, so that we can respect the preferred FPS value, which is specified by HU.
```

I think this line should be moved to the `Impact on Existing` code section of the proposal.